File: HaskellTree/BinarySearchTree.hs
{-|
Module      : BinarySearchTree

Ryan Forsyth

May 8, 2015
-}

module BinarySearchTree (
    Tree(None),
    inOrder,
    preOrder,
    postOrder,
    insert,
    insertMultiple,
    search,
) where


-- Constructors

data Tree a = None | Node a (Tree a) (Tree a)


-- Instances

instance (Eq a) => Eq (Tree a) where
    None == None = True
    Node n left right == Node n' left' right' = (n == n') &&
                                                (left == left') &&
                                                (right == right')
    _ == _ = False

instance (Show a) => Show (Tree a) where
    show = show . inOrder

instance Functor Tree where
    fmap f None = None
    fmap f (Node n left right) = Node (f n) (fmap f left) (fmap f right)


-- Tree Traversals
    
-- | The in-order traversal of the tree
inOrder :: Tree a -> [a]
inOrder None = []
inOrder (Node n left right) = (inOrder left) ++ [n] ++ (inOrder right)

-- | The pre-order traversal of the tree
preOrder :: Tree a -> [a]
preOrder None = []
preOrder (Node n left right) = [n] ++ (preOrder left) ++ (preOrder right)

-- | The post-order traversal of the tree
postOrder :: Tree a -> [a]
postOrder None = []
postOrder (Node n left right) = (postOrder left) ++ (postOrder right) ++ [n]


-- Insertion

-- | Insert an element into the tree
insert :: (Ord a) => Tree a ->  a -> Tree a
insert None element = Node element None None
insert (Node n left right) element =
    if (element < n)
    then Node n (insert left element) right
    else Node n left (insert right element)

-- | Insert multiple elements into the tree
insertMultiple :: (Ord a) => Tree a -> [a] -> Tree a
insertMultiple tree elements = foldl insert tree elements


-- Searching

-- | Search for an item in the tree
search :: (Ord a) => Tree a -> a -> Bool
search None _ = False
search (Node n left right) element
   | element < n  = search left element
   | element > n  = search right element
   | element == n = True
   | otherwise = error "Impossible"
    

File: OCamlTree/binarySearchTree.ml
(* Binary Search Tree
   Ryan Forsyth
   May 10, 2015 *)
   
type 'a tree =
   | Empty
   | Node of 'a * 'a tree * 'a tree


(* Tree Traversals *)

let rec inOrder (t : 'a tree) : 'a list =
   match t with
      Empty -> []
   |  Node (n, left, right) -> inOrder left @ [n] @ inOrder right
   
let rec preOrder (t : 'a tree) : 'a list =
   match t with
      Empty -> []
   |  Node (n, left, right) -> [n] @ preOrder left @ preOrder right
   
let rec postOrder (t : 'a tree) : 'a list =
   match t with
      Empty -> []
   |  Node (n, left, right) -> postOrder left @ postOrder right @ [n]


(* Insertion *)

let rec insert (t : 'a tree) (e : 'a) : 'a tree =
   match t with
      Empty -> Node (e, Empty, Empty)
   |  Node (n, left, right) ->
      if e < n
      then Node (n, (insert left e), right)
      else Node (n, left, (insert right e))

let insertMultiple (t : 'a tree) (es : 'a list) : 'a tree =
   List.fold_left insert t es


(* Searching *)

let rec search (t : 'a tree) (e: 'a) : bool =
   match t with
      Empty -> false
   |  Node (n, left, right) ->
      if e < n
      then search left e
      else if e > n
      then search right e
      else true
